src/middleware/messageStore.coffee

```diff
@@ -65,70 +69,119 @@ exports.storeTransaction = (ctx, done) ->
+      ctx.header['channel-id'] = tx.channelID
       return done null, tx
 
 exports.storeResponse = (ctx, done) ->
   logger.info 'Storing response for transaction: ' + ctx.transactionId
+  getNumRoutes ctx.authorisedChannel._id, (numRoutes) ->
+    status = null
+    if (500 <= ctx.response.status <= 599)
+      status = transactionStatus.FAILED
+    else
+      if (200 <= ctx.response.status <= 299)
+        console.log 'in Primary successful'
+        status = transactionStatus.SUCCESSFUL
+
+      # In all other cases mark as completed
+      if status is null or status is undefined
+        console.log 'in Primary Completed'
+        status = transactionStatus.COMPLETED
+
+    headers = copyMapWithEscapedReservedCharacters ctx.response.header
+
+    res =
+      status: ctx.response.status
+      headers: headers
+      body: if not ctx.response.body then "" else ctx.response.body.toString()
+      timestamp: ctx.response.timestamp
+
+
+    # check if channel response body is false and remove
+    if ctx.authorisedChannel.responseBody == false
+      # reset request body - primary route
+      res.body = ''
+
+      # reset request body - routes
+      if ctx.routes
+        for route in ctx.routes
+          route.response.body = ''
+
+
+    # assign new transactions status to ctx object
+    ctx.transactionStatus = status
+
+    update = { response: res, status: status }
+
+    if ctx.mediatorResponse
+      update.orchestrations = ctx.mediatorResponse.orchestrations if ctx.mediatorResponse.orchestrations
+      update.properties = ctx.mediatorResponse.properties if ctx.mediatorResponse.properties
+
+    transactions.Transaction.findOneAndUpdate { _id: ctx.transactionId }, update, (err, tx) ->
+      if err
+        logger.error 'Could not save response metadata for transaction: ' + ctx.transactionId + '. ' + err
+        return done err
+      if tx is undefined or tx is null
+        logger.error 'Could not find transaction: ' + ctx.transactionId
+        return done err
+      return done()
+
+exports.getNumRoutes = getNumRoutes = (channelID, callback) ->
+  Channel.findById channelID, (err, channel) ->
+    if channel?.routes?
+      callback channel.routes.length
+    else
+      callback 0
+
+exports.storeNonPrimaryResponse = (ctx, response, done) ->
+  #  get channel and determine number of routes in channel
+  if ctx.request.header?["channel-id"]?
+    getNumRoutes ctx.request.header["channel-id"], (numRoutes) ->
+    #  Get the current transaction and get the number of routes that have completed
+      do (numRoutes) ->
+        if ctx.request.header?["X-OpenHIM-TransactionID"]?
+          transactions.Transaction.findById ctx.request.header["X-OpenHIM-TransactionID"], (err,tx) ->
+            do (tx) ->
+              numRouteResps = tx.routes.length + 1
+              remainingRoutes = numRoutes - numRouteResps
+              logger.info "num remaining routes: #{remainingRoutes}"
+              isLastRoute = false
```

Comment:
Rather do the last route processing in another function and call it when the non primary route promises resolve. You have this already in the router.